OctoSPI and Memory Mapped Flash support

hydra opened the pull request:
This contains the commits from the #11110 to support:

* OctoSPI bus.  OctoSPI replaces QuadSPI on newer ST MCUs.  OctoSPI works in a very similar manner to QuadSPI in that you configure hardware registers for SPI commands/addresses/data/dummy parts for the transfer, then do the actual transfer.
* Memory Mapped flash support.  Memory mapped flash support allows the BF firmware to run directly from external flash, this allows flash-size-constrained MCUs like the H730 and H750 'value-line' boards.  This is builds on the existing EXST support where the code is compiled for, and flashed to, an external flash chip.
* Config in Memory Mapped flash.  In order for this to work, all the code that writes to the flash that the firmware is running from must be in RAM.  When Memory Mapped mode is disabled executing instructions from the external flash is no-longer allowed.  The basic process is 1) annotate methods with a linker attribute that causes them to be in RAM.  2) Disable memory mapped mode, 3) use traditional Quad/OctoSPI commands and flash drivers to write the config to sectors on the external flash chip, then 4) re-enable memory mapped mode.

Additionally this PR adds some other commits to:
* Fix issues in the NucleoH723 target.
* Adds a 'NucleoH725ZG_MMAPPED' target, this is done so a) there is an example of how it all fits together, b) there is CI visibility.

Also note:
* An additional H730 based target, the SPRacingH7RF is in PR #11110 which uses the H730 instead of the H723, but the implementation is identical.
* H723/H725/H730/H733/H735 are pin-compatible when it comes to the OctoSPI pins used in this PR.
* The code here is tested on the H730, H723 should work just fine, I'll update this once I've had time to hook up a suitable flash chip to one of the eval boards, I imagine the only issues would be in some of the early MCU startup code as while developing for the H730 I took care to gate MCU specific code around being compiled for the H730 so the commit that adds H723 support should cover the required changes.
* The commits are in a somewhat unusual order, this is due to the order of development of the OctoSPI support which was done on the H730, which required memory mapped mode support BEFORE flash and config-in-memory-mapped-flash support.
* Yes, I can squash some of the commits prior to merge.

Here's what the memory usage looks like:
```
Linking NUCLEOH723ZG_MMAPPED 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       16336 B        64 KB     24.93%
        DTCM_RAM:       33520 B       128 KB     25.57%
             RAM:      100896 B       320 KB     30.79%
          D2_RAM:          0 GB        32 KB      0.00%
          D3_RAM:          0 GB        16 KB      0.00%
     BACKUP_SRAM:          0 GB         4 KB      0.00%
       MEMORY_B1:          0 GB         0 GB
        OCTOSPI2:          0 GB       256 MB      0.00%
        OCTOSPI1:          0 GB       256 MB      0.00%
   OCTOSPI1_CODE:      525245 B    1048512 B     50.09%
       EXST_HASH:          64 B         64 B    100.00%
   text	   data	    bss	    dec	    hex	filename
 518125	   7180	 124736	 650041	  9eb39	./obj/main/betaflight_NUCLEOH723ZG_MMAPPED.elf
 ...
 Creating EXST HEX from patched EXST BIN ./obj/betaflight_4.4.0_NUCLEOH723ZG_MMAPPED.bin, VMA Adjust 0x90100000
 ```
 
Note the code is in `OCTOSPI1_CODE`, not `FLASH`.

If reviewers want to hop onto a video call with me I'm happy to oblige.  PM me on slack/discord to arrange, the more the merrier.

Update:

Testers can also use the `support/scripts/build_spracingh7rf.sh` script to build a target that will boot an an SPRacingH7RF.

```
Linking STM32H730
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       18760 B        64 KB     28.63%
        DTCM_RAM:       32752 B       128 KB     24.99%
             RAM:       98932 B       320 KB     30.19%
          D2_RAM:          0 GB        32 KB      0.00%
          D3_RAM:          0 GB        16 KB      0.00%
     BACKUP_SRAM:          0 GB         4 KB      0.00%
       MEMORY_B1:          0 GB         0 GB
        OCTOSPI2:          0 GB       256 MB      0.00%
        OCTOSPI1:          0 GB       256 MB      0.00%
   OCTOSPI1_CODE:      539111 B    1048512 B     51.42%
       EXST_HASH:          64 B         64 B    100.00%
```

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

ledvinap commented:
What about blackbox? Is it disabled in memory mapped mode?

hydra commented:
> What about blackbox? Is it disabled in memory mapped mode?

It's true that you can't use it. However I have another target which has an SPI flash chip *and* a OctoSPI connected flash so you can't gate availability of BB logging to flash on memory mapped mode.  Similarly the H7RF has BB on SD card + MM Flash. Currently targets can specify to use BB logging on sdcard or flash.

hydra commented:
@ledvinap I pushed a commit so that the target definiton excludes FlashFS when using Memory Mapped flash.  I'd forgotten that I had done that in my other targets.  Thanks for the review.

@haslinghuis Ready for approval/review again.

ledvinap commented on src/main/drivers/flash.c:
```diff
@@ -47,13 +51,142 @@
+        }
+#endif
+
+        for (uint8_t offset = 0; offset <= 1 && !detected; offset++) {
+
+            uint32_t jedecID = (readIdResponse[offset + 0] << 16) | (readIdResponse[offset + 1] << 8) | (readIdResponse[offset + 2]);
+
+            if (offset == 0) {
+#if defined(USE_FLASH_W25Q128FV)
+                if (!detected && w25q128fv_identify(&flashDevice, jedecID)) {
+                    detected = true;
+                }
+#endif
+            }
+
+            if (offset == 1) {
+#ifdef USE_OCTOSPI_EXPERIMENTAL
+                if (!memoryMappedModeEnabledOnBoot) {
+                    // These flash chips DO NOT support memory mapped mode; suitable flash read commands must be available.
+#if defined(USE_FLASH_W25N01G)
+                    if (!detected && w25n01g_identify(&flashDevice, jedecID)) {
+                        detected = true;
+                    }
+#endif
+#if defined(USE_FLASH_W25M02G)
+                    if (!detected && w25m_identify(&flashDevice, jedecID)) {
+                        detected = true;
+                    }
+#endif
+                }
+#endif
+            }
```
```suggestion
            uint32_t jedecID = (readIdResponse[0] << 24) | (readIdResponse[1] << 16) | (readIdResponse[2] << 8) | (readIdResponse[3]);
#if defined(USE_FLASH_W25Q128FV)
                if (w25q128fv_identify(&flashDevice, jedecID >> 8)) {
                    detected = true;
                    break;
                }
#endif
#ifdef USE_OCTOSPI_EXPERIMENTAL
                if (!memoryMappedModeEnabledOnBoot) {
                    // These flash chips DO NOT support memory mapped mode; suitable flash read commands must be available.
#if defined(USE_FLASH_W25N01G)
                    if (w25n01g_identify(&flashDevice, jedecID & 0xffffff)) {
                        detected = true;
                        break;
                    }
#endif
#if defined(USE_FLASH_W25M02G)
                    if (w25m_identify(&flashDevice, jedecID & 0xffffff)) {
                        detected = true;
                        break;
                    }
#endif
                }
#endif
            }
```

and possibly use macro/function for jedecID alignment

hydra replied:
this follows the pattern of the QSPI init code.  A later commit can address both at the same time, but for now they are consistent.

ledvinap commented on src/main/drivers/flash.c:
```diff
@@ -47,13 +51,142 @@
+            continue;
+        }
+
+#ifdef USE_OCTOSPI_EXPERIMENTAL
+        if (!memoryMappedModeEnabledOnBoot) {
+            octoSpiSetDivisor(instance, OCTOSPI_CLOCK_ULTRAFAST);
+        }
+#endif
```
Clock setting may be outside do { } while();

ledvinap commented on src/main/drivers/flash.c:
```diff
@@ -208,39 +335,65 @@ static bool flashSpiInit(const flashConfig_t *flashConfig)
 
-    return false;
+#ifdef USE_FLASH_OCTOSPI
+    bool useOctoSpi = (OCTOSPI_CFG_TO_DEV(flashConfig->octoSpiDevice) != OCTOSPIINVALID);
```
(default value is designed to allow `if (flashConfig->octoSpiDevice)` )

ledvinap commented on src/main/drivers/flash.c:
```diff
@@ -208,39 +335,65 @@ static bool flashSpiInit(const flashConfig_t *flashConfig)
-#ifdef USE_QUADSPI
+#ifdef USE_FLASH_QUADSPI
     bool useQuadSpi = (QUADSPI_CFG_TO_DEV(flashConfig->quadSpiDevice) != QUADSPIINVALID);
     if (useQuadSpi) {
```
Old code did return on first flash found
```suggestion
    if (!haveFlash && useQuadSpi) {
```
Maybe use `do { .. } while(0);` and break on first `haveFlash`

ledvinap commented on src/main/drivers/flash.c:
```diff
@@ -208,39 +335,65 @@ static bool flashSpiInit(const flashConfig_t *flashConfig)
+        }
+#endif
+
+        flashDevice.vTable->configure(&flashDevice, configurationFlags);
```
None: This was not called for spi/quad flash before

ledvinap commented on src/main/drivers/flash_impl.h:
```diff
@@ -32,20 +56,30 @@ struct flashVTable_s;
 typedef struct flashDeviceIO_s {
     union {
+    #ifdef USE_FLASH_SPI
         extDevice_t *dev; // Device interface dependent handle (spi/i2c)
```
no I2C now ?

hydra replied:
this PR hasn't changed this, see the initial QUADSPI commit (dcd138ae20672a3305dcf3aa1a19cab8c955dbad) from a few years ago which added `FLASHIO_SPI`.  There's never been `FLASHIO_I2C`.  If flash by I2C support is added then the PR to add it can address this.

ledvinap commented on src/main/drivers/flash_w25q128fv.c:
```diff
@@ -115,45 +119,71 @@ w25q128fvState_t w25q128fvState = { 0 };
-    uint32_t now = HAL_GetTick();
-    fdevice->timeoutAt = now + timeoutMillis;
+    timeMs_t nowMs = microsISR() / 1000;
+    fdevice->timeoutAt = nowMs + timeoutMillis;
```
`micros() / 1000` will behave terribly on overflow (you can set timeout to value outside timer/1000 range)
Better multiply by 1000 and use micros

And is this code really called only from ISR ?

hydra replied:
The wait/delay functions must work with interrupts disabled, so microsISR() is needed as millis() requires systick to be enabled.

See git blame on those lines which reveals: 2125ce14633b29cd4721735eea7e62bc2322f7d7

ledvinap commented on src/main/drivers/flash_w25q128fv.c:
```diff
@@ -115,45 +119,71 @@ w25q128fvState_t w25q128fvState = { 0 };
-static uint8_t w25q128fv_readRegister(flashDeviceIO_t *io, uint8_t command)
+MMFLASH_CODE static uint8_t w25q128fv_readRegister(flashDeviceIO_t *io, uint8_t command)
 {
+    uint8_t in[1] = { 0 };
```
```suggestion
    uint8_t in[W25Q128FV_STATUS_REGISTER_BITS / 8] = { 0 };
```

hydra replied:
Done, also the same issue was in the W25N driver where there was also a typo which was using `W28N...`, 2 commits added.

ledvinap commented on src/main/drivers/flash_w25q128fv.c:
```diff
@@ -115,45 +119,71 @@ w25q128fvState_t w25q128fvState = { 0 };
+MMFLASH_CODE static uint8_t w25q128fv_readRegister(flashDeviceIO_t *io, uint8_t command)
 {
+    uint8_t in[1] = { 0 };
+#ifdef USE_QUADSPI
```
```suggestion
#if defined(USE_QUADSPI)
```
to match #elif style

hydra replied:
good catch, fixed.

ledvinap commented on src/main/drivers/flash_w25q128fv.c:
```diff
@@ -221,23 +251,23 @@ static bool w25q128fv_isWritable(flashDevice_t *fdevice)
-    uint32_t now = HAL_GetTick();
-    if (cmp32(now, fdevice->timeoutAt) >= 0) {
+    uint32_t nowMs = microsISR() / 1000;
+    if (cmp32(nowMs, fdevice->timeoutAt) >= 0) {
```
This is the problem from above - cmp32() works only  with 32bit overflow

hydra replied:
Ahh yes, suggestions on what to do given that the timeout/delay functions need to work without systick ticking up the millis counter via `SysTick_Handler` then?

hydra replied:
since there's no suggestions for a month, and the code has been running on H7RF targets for over a year with zero reported problems we can look to improve this in the future once a suggestion/solution presents itself.

hydra commented:
@ledvinap thanks for the review. I'll dedicate some time soon and update this PR based on your comments and suggestions.

hydra commented:
I'll rebase this on `master` and address the comments now that #12093 is merged.

hydra commented:
Rebased on `master`, other comments not yet addressed.

Use `support/scripts/build_nucleoh723zg_mmapped.sh` to build an H723 target with MMFLASH/OCTOSPI support.

Build log excerpt:
```
Linking STM32H723
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       17560 B        64 KB     26.79%
        DTCM_RAM:       32432 B       128 KB     24.74%
             RAM:      101748 B       320 KB     31.05%
          D2_RAM:          0 GB        32 KB      0.00%
          D3_RAM:          0 GB        16 KB      0.00%
     BACKUP_SRAM:          0 GB         4 KB      0.00%
       MEMORY_B1:          0 GB         0 GB
        OCTOSPI2:          0 GB       256 MB      0.00%
        OCTOSPI1:          0 GB       256 MB      0.00%
   OCTOSPI1_CODE:      570615 B    1048512 B     54.42%
       EXST_HASH:          64 B         64 B    100.00%
   text    data     bss     dec     hex filename
 561379    9296  122364  693039   a932f ./obj/main/betaflight_STM32H723.elf
Creating BIN (without checksum) ./obj/main/betaflight_STM32H723_UNPATCHED.bin
Creating EXST ./obj/betaflight_4.4.0_STM32H723.bin
1024+0 records in
2048+0 records out
1048576 bytes (1.0 MB, 1.0 MiB) copied, 0.0092153 s, 114 MB/s
2048+0 records in
2048+0 records out
1048576 bytes (1.0 MB, 1.0 MiB) copied, 0.0123618 s, 84.8 MB/s
Generating MD5 hash of binary
Patching MD5 hash into binary
000ffff0: 8209 a4bc 1e5f f934 a49d 2678 edd2 7d65  ....._.4..&x..}e
Extracting HASH section from unpatched EXST elf ./obj/main/betaflight_STM32H723.elf
./tools/gcc-arm-none-eabi-10.3-2021.10/bin/arm-none-eabi-objcopy ./obj/main/betaflight_STM32H723.elf ./obj/main/betaflight_STM32H723_EXST.elf.tmp --dump-section .exst_hash=./obj/main/STM32H723/exst_hash_section.bin -j .exst_hash
rm ./obj/main/betaflight_STM32H723_EXST.elf.tmp
Patching MD5 hash into HASH section
Patching updated HASH section into ./obj/main/betaflight_STM32H723_EXST.elf
./tools/gcc-arm-none-eabi-10.3-2021.10/bin/arm-none-eabi-objcopy ./obj/main/betaflight_STM32H723.elf ./obj/main/betaflight_STM32H723_EXST.elf --update-section .exst_hash=./obj/main/STM32H723/exst_hash_section.bin
There are 36 section headers, starting at offset 0xa157ec:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .isr_vector       PROGBITS        90100000 010000 0002cc 00   A  0   0  1
  [ 2] .ram_isr_vector   NOBITS          20000000 010000 0002cc 00  WA  0   0  1
  [ 3] .text             PROGBITS        901002d0 0102d0 0832ec 00  AX  0   0  8
  [ 4] .tcm_code         PROGBITS        00000000 0a0000 004498 00  AX  0   0  8
  [ 5] .ARM              ARM_EXIDX       90187a54 0a7a54 000008 00  AL  3   0  4
  [ 6] .pg_registry      PROGBITS        90187a5c 0a7a5c 000a64 00   A  0   0  4
  [ 7] .pg_resetdata     PROGBITS        901884c0 0a84c0 00023f 00   A  0   0  4
  [ 8] .data             PROGBITS        200002cc 0b02cc 001e3c 00  WA  0   0  4
  [ 9] .ram_code         PROGBITS        24000000 0c0000 0009e8 00  AX  0   0  8
  [10] .bss              NOBITS          240009e8 0d09e8 00fde0 00  WA  0   0  8
  [11] .sram2            PROGBITS        240107c8 0e0000 000000 00   W  0   0  1
  [12] .fastram_data     PROGBITS        20002108 0c2108 000034 00  WA  0   0  4
  [13] .fastram_bss      NOBITS          2000213c 00213c 005574 00  WA  0   0  4
  [14] .dmaram_data      PROGBITS        240107e0 0d07e0 0005a0 00  WA  0   0 32
  [15] .dmaram_bss       NOBITS          24010d80 0d0d80 001680 00  WA  0   0 32
  [16] .DMA_RAM          NOBITS          24012400 0d2400 005ee8 00  WA  0   0 32
  [17] .DMA_RW_D2        PROGBITS        30000000 0e0000 000000 00   W  0   0  1
  [18] .DMA_RW_AXI       NOBITS          24018300 008300 000a74 00  WA  0   0 32
  [19] .persistent_data  PROGBITS        24018d74 0e0000 000000 00   W  0   0  1
  [20] ._user_heap_stack NOBITS          200076b0 010000 000800 00  WA  0   0  1
  [21] .ARM.attributes   ARM_ATTRIBUTES  00000000 0e0000 000035 00      0   0  1
  [22] .exst_hash        PROGBITS        901fffc0 0dffc0 000040 00  WA  0   0  1
  [23] .debug_line       PROGBITS        00000000 0e0035 1c4d1f 00      0   0  1
  [24] .debug_info       PROGBITS        00000000 2a4d54 25da81 00      0   0  1
  [25] .debug_abbrev     PROGBITS        00000000 5027d5 02e4c9 00      0   0  1
  [26] .debug_aranges    PROGBITS        00000000 530ca0 000480 00      0   0  8
  [27] .debug_str        PROGBITS        00000000 531120 22b66d 01  MS  0   0  1
  [28] .debug_ranges     PROGBITS        00000000 75c790 05d9b8 00      0   0  8
  [29] .debug_loc        PROGBITS        00000000 7ba148 129162 00      0   0  1
  [30] .comment          PROGBITS        00000000 8e32aa 000049 01  MS  0   0  1
  [31] .debug_frame      PROGBITS        00000000 8e32f4 00ed10 00      0   0  4
  [32] .debug_macro      PROGBITS        00000000 8f2004 0b04ab 00      0   0  1
  [33] .symtab           SYMTAB          00000000 9a24b0 057060 10     34 19456  4
  [34] .strtab           STRTAB          00000000 9f9510 01c149 00      0   0  1
  [35] .shstrtab         STRTAB          00000000 a15659 000191 00      0   0  1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  y (purecode), p (processor specific)

Elf file type is EXEC (Executable file)
Entry point 0x9016f851
There are 13 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x20000000 0x20000000 0x00000 0x07eb0 RW  0x10000
  LOAD           0x008300 0x24018300 0x24018300 0x00000 0x00a74 RW  0x10000
  LOAD           0x010000 0x90100000 0x90100000 0x835bc 0x835bc R E 0x10000
  LOAD           0x0a0000 0x00000000 0x901835bc 0x04498 0x04498 R E 0x10000
  LOAD           0x0a7a54 0x90187a54 0x90187a54 0x00cab 0x00cab R   0x10000
  LOAD           0x0b02cc 0x200002cc 0x901886ff 0x01e3c 0x01e3c RW  0x10000
  LOAD           0x0c0000 0x24000000 0x9018a53b 0x009e8 0x009e8 R E 0x10000
  LOAD           0x0c2108 0x20002108 0x9018af23 0x00034 0x00034 RW  0x10000
  LOAD           0x0009e8 0x240009e8 0x9018af23 0x00000 0x0fde0 RW  0x10000
  LOAD           0x00213c 0x2000213c 0x2000213c 0x00000 0x05574 RW  0x10000
  LOAD           0x0d07e0 0x240107e0 0x9018af57 0x005a0 0x01c20 RW  0x10000
  LOAD           0x002400 0x24012400 0x9018b4f7 0x00000 0x05ee8 RW  0x10000
  LOAD           0x0dffc0 0x901fffc0 0x901fffc0 0x00040 0x00040 RW  0x10000

 Section to Segment mapping:
  Segment Sections...
   00     .ram_isr_vector .fastram_bss ._user_heap_stack
   01     .DMA_RW_AXI
   02     .isr_vector .text
   03     .tcm_code
   04     .ARM .pg_registry .pg_resetdata
   05     .data
   06     .ram_code
   07     .fastram_data
   08     .bss
   09     .fastram_bss
   10     .dmaram_data .dmaram_bss
   11     .DMA_RAM
   12     .exst_hash
Creating EXST HEX from patched EXST BIN ./obj/betaflight_4.4.0_STM32H723.bin, VMA Adjust 0x90100000
make[1]: Leaving directory '/cygdrive/d/Users/Hydra/Documents/dev/projects/betaflight/betaflight'
```

hydra commented on src/main/drivers/bus_octospi.c:
```diff
@@ -0,0 +1,119 @@
+        const octoSpiHardware_t *hw = &octoSpiHardware[hwindex];
+
+        OCTOSPIDevice device = hw->device;
+        octoSpiDevice_t *pDev = &octoSpiDevice[device];
+
+        pDev->dev = hw->reg;
```
no, `pDev->dev` - each `octoSpiDevice` needs it's `device` member initialised from the corresponding `octoSpiHardware`.

hydra commented on src/main/drivers/flash.c:
```diff
@@ -47,13 +51,142 @@
+{
+    bool detected = false;
+
+    enum { TRY_1LINE = 0, TRY_4LINE, BAIL};
+    int phase = TRY_1LINE;
```
done

hydra commented on src/main/drivers/bus_octospi_stm32h7xx.c:
```diff
@@ -0,0 +1,854 @@
+#define OSPI_SIOO_INST_ONLY_FIRST_CMD    ((uint32_t)OCTOSPI_CCR_SIOO)
+
+MMFLASH_CODE_NOINLINE static void Error_Handler(void) {
+    while (1)
+    {
```
done

hydra commented:
> What about blackbox? Is it disabled in memory mapped mode?

on mmflash flash, yes, though I do have an H730 target which has OCTOSPI mmflash and a second flash for config storage and blackbox logging, the SPRacingH7EF.  See the SPRacing BF repo's 'h7rf wip' branches.  It works great.

hydra commented:
rebased on `master`

hydra commented on src/main/config/config_streamer.c:
```diff
@@ -541,11 +541,13 @@ int config_streamer_finish(config_streamer_t *c)
 #elif defined(CONFIG_IN_EXTERNAL_FLASH)
         flashFlush();
+#elif defined(CONFIG_IN_MEMORY_MAPPED_FLASH)
+        void saveEEPROMToMemoryMappedFlash(void); // forward declaration to avoid circular dependency between config_streamer / config_eeprom
```
agreed, but this follows the pattern used when `CONFIG_IN_SDCARD` was added, see a few lines up.  Both can be fixed as part of a new PR to address the headers for both in a consistent way.

hydra commented:
OP updated with a note regarding the availability of `support/scripts/build_spracingh7rf.sh` which can be used to build an H730 unified target that will boot on the SPRacingH7RF.

hydra commented:
```
# status
MCU H730 Clock=520MHz, Vref=3.31V, Core temp=42degC
Stack size: 2048, Stack address: 0x20020000
Configuration: UNCONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:0, I2C:0
Gyros detected: locked

OSD: MSP (30 x 13)
System Uptime: 7 seconds, Current Time: 2023-01-28T21:20:25.523+00:00
CPU:0%, cycle time: 0, GYRO rate: 0, RX rate: 0, System rate: 9
Voltage: 63 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
SD card: None inserted
Arming disable flags: NOGYRO BOOTGRACE CLI MSP

# version
# Betaflight / STM32H730 (SP7R) 4.4.0 Jan 28 2023 / 21:17:30 (d78f0c9183) MSP API: 1.45
```

Binary:

[SPRacingH7RF-octo-spi-test-1.zip](https://github.com/betaflight/betaflight/files/10528316/SPRacingH7RF-octo-spi-test-1.zip)

hydra commented:
Rebased again, conflicts in #12221 necessitated a new PR #12252.

hydra commented:
Updated binaries for testing:

[SPRacingH7RF-octo-spi-test-3.zip](https://github.com/betaflight/betaflight/files/10530615/SPRacingH7RF-octo-spi-test-3.zip)

Output:

```
# Betaflight / STM32H730 (SP7R) 4.5.0 Jan 29 2023 / 19:08:00 (a7d4590b31) MSP API: 1.45
# board: manufacturer_id: SPRO, board_name: SPRACINGH7EXTREME

# status
MCU H730 Clock=520MHz, Vref=3.31V, Core temp=44degC
Stack size: 2048, Stack address: 0x20020000
Configuration: CONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked
GYRO=ICM42605, ACC=ICM42605, BARO=BMP388
OSD: NONE (104 x 34)
System Uptime: 24 seconds, Current Time: 2023-01-29T19:11:42.052+00:00
CPU:30%, cycle time: 124, GYRO rate: 8064, RX rate: 15, System rate: 9
Voltage: 65 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 9
SD card: None inserted
Arming disable flags: RXLOSS CLI MSP MOTOR_PROTO

```

Diff, for H7RF.
```
# diff

# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Jan 29 2023 / 19:08:00 (a7d4590b31) MSP API: 1.45

# start the command batch
batch start

board_name SPRACINGH7EXTREME
manufacturer_id SPRO

# resources
resource BEEPER 1 E04
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A06
resource MOTOR 4 A07
resource MOTOR 5 A00
resource MOTOR 6 A01
resource MOTOR 7 A02
resource MOTOR 8 A03
resource LED_STRIP 1 B07
resource SERIAL_TX 2 D05
resource SERIAL_TX 3 D08
resource SERIAL_TX 4 D01
resource SERIAL_TX 8 E01
resource SERIAL_RX 2 D06
resource SERIAL_RX 3 D09
resource SERIAL_RX 4 D00
resource SERIAL_RX 5 B13
resource SERIAL_RX 8 E00
resource LED 1 E05
resource LED 2 E06
resource GYRO_EXTI 1 D15
resource GYRO_CS 1 A15

# feature
feature TELEMETRY
feature LED_STRIP

# master
set mag_bustype = I2C
set mag_i2c_device = 1
set baro_bustype = I2C
set baro_i2c_device = 2
set pinio_box = 40,255,255,255
set gyro_1_spibus = 6
set gyro_1_sensor_align = CW270FLIP
set gyro_1_align_pitch = 1800
set gyro_1_align_yaw = 2700

profile 0

rateprofile 0

# end the command batch
batch end

```

blockfeed commented:
Works for me!
```

# status
MCU H730 Clock=520MHz, Vref=3.33V, Core temp=51degC
Stack size: 2048, Stack address: 0x20020000
Configuration: CONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked
GYRO=ICM42688P, ACC=ICM42688P, BARO=BMP388
OSD: NONE (104 x 34)
System Uptime: 80 seconds, Current Time: 2023-01-30T05:59:30.533+00:00
CPU:30%, cycle time: 125, GYRO rate: 8000, RX rate: 15, System rate: 9
Voltage: 125 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 9
SD card: None inserted
Arming disable flags: RXLOSS CLI MSP MOTOR_PROTO
```

haslinghuis commented:
@hydra please rebase

hydra commented:
[SPRacingH7RF-octo-spi-test-4.zip](https://github.com/betaflight/betaflight/files/10557173/SPRacingH7RF-octo-spi-test-4.zip)

```
# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  1 2023 / 12:55:21 (norevision) MSP API: 1.45
# board: manufacturer_id: SPRO, board_name: SPRACINGH7EXTREME

# status
MCU H730 Clock=520MHz, Vref=3.31V, Core temp=48degC
Stack size: 2048, Stack address: 0x20020000
Configuration: CONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked
GYRO=ICM42605, ACC=ICM42605, BARO=BMP388
OSD: NONE (104 x 34)
System Uptime: 109 seconds, Current Time: 2023-02-01T13:00:51.992+00:00
CPU:34%, cycle time: 124, GYRO rate: 8064, RX rate: 15, System rate: 9
Voltage: 1199 * 0.01V (3S battery - OK)
I2C Errors: 9
SD card: None inserted
Arming disable flags: RXLOSS CLI MSP MOTOR_PROTO

# diff

# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  1 2023 / 12:55:21 (norevision) MSP API: 1.45

# start the command batch
batch start

board_name SPRACINGH7EXTREME
manufacturer_id SPRO

# resources
resource BEEPER 1 E04
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A06
resource MOTOR 4 A07
resource MOTOR 5 A00
resource MOTOR 6 A01
resource MOTOR 7 A02
resource MOTOR 8 A03
resource LED_STRIP 1 B07
resource SERIAL_TX 2 D05
resource SERIAL_TX 3 D08
resource SERIAL_TX 4 D01
resource SERIAL_TX 8 E01
resource SERIAL_RX 2 D06
resource SERIAL_RX 3 D09
resource SERIAL_RX 4 D00
resource SERIAL_RX 5 B13
resource SERIAL_RX 8 E00
resource LED 1 E05
resource LED 2 E06
resource GYRO_EXTI 1 D15
resource GYRO_CS 1 A15

# feature
feature TELEMETRY
feature LED_STRIP

# master
set mag_bustype = I2C
set mag_i2c_device = 1
set baro_bustype = I2C
set baro_i2c_device = 2
set pinio_box = 40,255,255,255
set gyro_1_spibus = 6
set gyro_1_sensor_align = CW270FLIP
set gyro_1_align_pitch = 1800
set gyro_1_align_yaw = 2700

profile 0

rateprofile 0

# end the command batch
batch end
```

@haslinghuis done

hydra commented:
rebased again and squashed the rebase fixes into the correct commits...

[SPRacingH7RF-octo-spi-test-5.zip](https://github.com/betaflight/betaflight/files/10557267/SPRacingH7RF-octo-spi-test-5.zip)

```
# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  1 2023 / 13:09:23 (eca29a29ae) MSP API: 1.45
# board: manufacturer_id: SPRO, board_name: SPRACINGH7EXTREME

# status
MCU H730 Clock=520MHz, Vref=3.31V, Core temp=51degC
Stack size: 2048, Stack address: 0x20020000
Configuration: CONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked
GYRO=ICM42605, ACC=ICM42605, BARO=BMP388
OSD: NONE (104 x 34)
System Uptime: 10 seconds, Current Time: 2023-02-01T13:12:23.530+00:00
CPU:34%, cycle time: 124, GYRO rate: 8064, RX rate: 15, System rate: 9
Voltage: 1198 * 0.01V (3S battery - OK)
I2C Errors: 9
SD card: None inserted
Arming disable flags: RXLOSS CLI MSP MOTOR_PROTO

# diff

# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  1 2023 / 13:09:23 (eca29a29ae) MSP API: 1.45

# start the command batch
batch start

board_name SPRACINGH7EXTREME
manufacturer_id SPRO

# resources
resource BEEPER 1 E04
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A06
resource MOTOR 4 A07
resource MOTOR 5 A00
resource MOTOR 6 A01
resource MOTOR 7 A02
resource MOTOR 8 A03
resource LED_STRIP 1 B07
resource SERIAL_TX 2 D05
resource SERIAL_TX 3 D08
resource SERIAL_TX 4 D01
resource SERIAL_TX 8 E01
resource SERIAL_RX 2 D06
resource SERIAL_RX 3 D09
resource SERIAL_RX 4 D00
resource SERIAL_RX 5 B13
resource SERIAL_RX 8 E00
resource LED 1 E05
resource LED 2 E06
resource GYRO_EXTI 1 D15
resource GYRO_CS 1 A15

# feature
feature TELEMETRY
feature LED_STRIP

# master
set mag_bustype = I2C
set mag_i2c_device = 1
set baro_bustype = I2C
set baro_i2c_device = 2
set pinio_box = 40,255,255,255
set gyro_1_spibus = 6
set gyro_1_sensor_align = CW270FLIP
set gyro_1_align_pitch = 1800
set gyro_1_align_yaw = 2700

profile 0

rateprofile 0

# end the command batch
batch end
```

blckmn reviewed (changes requested):
Please break this PR into respective component PRs

1. For introducing octospi 
2. Memory mapped support
3. Etc

hydra commented:
Here's more reasons why OctoSPI/QuadSPI is better for flash.

![image](https://user-images.githubusercontent.com/57075/216200207-6824f5b3-6ff8-457b-b545-77ec05ce5a22.png)

* Bus width (64 bits D1, vs 32bits for D2 SPI on APB2).
* Less Bus contention.  With gyros, flash and OSD on SPI the gyros has to wait for the bus to be free before it can be used to read gyro.
* Different DMA controller - DMA1/2 can be used for gyro while MDMA is uses for OctoSPI.

hydra commented:
> Please break this PR into respective component PRs
> 
>     1. For introducing octospi
> 
>     2. Memory mapped support
> 
>     3. Etc

@blckmn the current octospi support is ONLY for memory mapped flash, it thus cannot be tested without the memory mapped flash changes as the BF code won't boot as it has no-where to load/store it's config from.

Additional work can come after the base support for OctoSPI and a flash chip that can use OctoSPI by anyone who has time to write all the OctoSPI initialisation code.

David-OConnor commented:
Overall, this content being added to Betaflight is inevitable. The MCUs listed represent the most-capable STM32s available, and their extra performance will become more relevant as BF features are added and expanded. If pins are available, there is no reason *not* to use QuadSPI over normal SPI for external flash. Likewise if you have one of the MCUs listed here, OctoSPI is the clear choice. These setups require more pins, but this isn't a factor for the BGA packages, which I think will become the standard going forward due to a combination of compact size, and high pin count. (And are already common on existing H7-based FCs).

I have a design that will benefit from this capability, and as soon as I'm able to procure MCUs, will test the hardware, and this PR.

hydra commented:
rebased on `master`, again.

new binaries for testing, again:

[SPRacingH7RF-octo-spi-test-6.zip](https://github.com/betaflight/betaflight/files/10578310/SPRacingH7RF-octo-spi-test-6.zip)

```
# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  3 2023 / 11:09:34 (8a986b6fcb) MSP API: 1.45
###ERROR IN version: NO CONFIG FOUND#### board: manufacturer_id: SPRO, board_name: SPRACINGH7RF

# status
MCU H730 Clock=520MHz, Vref=3.31V, Core temp=41degC
Stack size: 2048, Stack address: 0x20020000
Configuration: CONFIGURED, size: 4138, max available: 65536
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked
GYRO=ICM42605, ACC=ICM42605, BARO=BMP388
OSD: NONE (104 x 34)
System Uptime: 75 seconds, Current Time: 2023-02-03T11:13:37.459+00:00
CPU:26%, cycle time: 125, GYRO rate: 8000, RX rate: 15, System rate: 9
Voltage: 59 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 9
SD card: None inserted
Arming disable flags: RXLOSS CLI MSP MOTOR_PROTO

# diff
###WARNING: NO CUSTOM DEFAULTS FOUND###

# version
# Betaflight / STM32H730 (SP7R) 4.5.0 Feb  3 2023 / 11:09:34 (8a986b6fcb) MSP API: 1.45
###ERROR IN diff: NO CONFIG FOUND###
# start the command batch
batch start

board_name SPRACINGH7RF
manufacturer_id SPRO

# resources
resource BEEPER 1 E04
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A06
resource MOTOR 4 A07
resource MOTOR 5 A00
resource MOTOR 6 A01
resource MOTOR 7 A02
resource MOTOR 8 A03
resource LED_STRIP 1 B07
resource SERIAL_TX 2 D05
resource SERIAL_TX 3 D08
resource SERIAL_TX 4 D01
resource SERIAL_TX 8 E01
resource SERIAL_RX 2 D06
resource SERIAL_RX 3 D09
resource SERIAL_RX 4 D00
resource SERIAL_RX 5 B13
resource SERIAL_RX 8 E00
resource LED 1 E05
resource LED 2 E06
resource GYRO_EXTI 1 D15
resource GYRO_CS 1 A15

# feature
feature TELEMETRY
feature LED_STRIP

# master
set mag_bustype = I2C
set mag_i2c_device = 1
set baro_bustype = I2C
set baro_i2c_device = 2
set pinio_box = 40,255,255,255
set gyro_1_spibus = 6
set gyro_1_sensor_align = CW270FLIP
set gyro_1_align_pitch = 1800
set gyro_1_align_yaw = 2700

profile 0

rateprofile 0

# end the command batch
batch end
```
